### ISSUE TITLE
Is the output of FlashAttention completely identical to that of vanilla attention?

### ISSUE THREAD

**Issue by sunsmarterjie** [Dao-AILab/flash-attention#1445]
I validated the outputs of both using the code below and found that the results are not exactly the same. Is this normal? What steps should I take to ensure both outputs are identical? Thank you!

Code:
'''
    import torch
    import math
    import torch.nn.functional as F
    from flash_attn import flash_attn_func

    def custom_attention(q, k, v, causal=False):
        score = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(q.size(-1))
        if causal:
            mask = torch.triu(torch.ones(score.shape[-2], score.shape[-1]), diagonal=1)
            mask = mask.masked_fill(mask==1, torch.finfo(q.dtype).min)
            mask = mask.to(q.device, q.dtype)
            score = score + mask
        attn = F.softmax(score, dim=-1)
        o = torch.matmul(attn, v)
        return o

    def flash_attention(q, k, v, causal=False):
        o = flash_attn_func(q, k, v)
        return o

    # 生成输入数据
    q = torch.randn(32, 4, 256, 64, dtype=torch.float16).cuda()
    k = torch.randn(32, 4, 256, 64, dtype=torch.float16).cuda()
    v = torch.randn(32, 4, 256, 64, dtype=torch.float16).cuda()

    # 计算普通 Attention
    custom_attn_output = custom_attention(q, k, v)
    print(custom_attn_output.mean(), custom_attn_output.sum())

    # 计算 Flash Attention
    flash_attn_output = flash_attention(q, k, v)
    print(flash_attn_output.mean(), flash_attn_output.sum())
'''

**Comment by tridao** [comment]
For others with the same question: floating point math is not associative, so one should never expect identical results for two different ways of computing the same quantity:
```python
In [1]: import torch

In [2]: a = torch.randn(1000, device='cuda', dtype=torch.bfloat16)

In [3]: b = torch.randn(1000, device='cuda', dtype=torch.bfloat16)

In [4]: c = torch.randn(1000, device='cuda', dtype=torch.bfloat16)

In [5]: torch.equal(a + b + c, b + c + a)
Out[5]: False
```

**Comment by XueniLuo** [comment]
> For others with the same question: floating point math is not associative, so one should never expect identical results for two different ways of computing the same quantity:
> 
> In [1]: import torch
> 
> In [2]: a = torch.randn(1000, device='cuda', dtype=torch.bfloat16)
> 
> In [3]: b = torch.randn(1000, device='cuda', dtype=torch.bfloat16)
> 
> In [4]: c = torch.randn(1000, device='cuda', dtype=torch.bfloat16)
> 
> In [5]: torch.equal(a + b + c, b + c + a)
> Out[5]: False

Hi, tridao. 

I've read some blogs quoted flash attention as " its outputs are the same as in the “vanilla” attention mechanism."
